pkg/apis/admissionregistration/types.go

```diff
@@ -115,6 +115,281 @@ const (
+	// Standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds
+	// +optional
+	metav1.ListMeta
+	// List of ValidatingAdmissionPolicy.
+	Items []ValidatingAdmissionPolicy
+}
+
+// ValidatingAdmissionPolicySpec is the specification of the desired behavior of the AdmissionPolicy.
+type ValidatingAdmissionPolicySpec struct {
+	// ParamKind specifies the kind of resources used to parameterize this policy.
+	// If absent, there are no parameters for this policy and the param CEL variable will not be provided to validation expressions.
+	// If ParamKind refers to a non-existent kind, this policy definition is mis-configured and the FailurePolicy is applied.
+	// +optional
+	ParamKind *ParamKind
+
+	// MatchConstraints specifies what resources this policy is designed to validate.
+	// The AdmissionPolicy cares about a request if it matches _any_ Constraint.
+	// However, in order to prevent clusters from being put into an unstable state that cannot be recovered from via the API
+	// ValidatingAdmissionPolicy cannot match ValidatingAdmissionPolicy and ValidatingAdmissionPolicyBinding.
+	// Required.
+	MatchConstraints *MatchResources
+
+	// Validations contain CEL expressions which is used to apply the validation.
+	// A minimum of one validation is required for a policy definition.
+	// Required.
+	Validations []Validation
+
+	// FailurePolicy defines how to handle failures for the admission policy.
+	// Failures can occur from invalid policy definitions or bindings, as well as validation errors from CEL expressions.
+	// A policy is invalid if spec.paramSource refers to a non-existent Kind.
+	// A binding is invalid if spec.paramName refers to a non-existent custom resource.
+	// allowed values are Ignore or Fail. Defaults to Fail.
+	// +optional
+	FailurePolicy *FailurePolicyType
+}
+
+// ParamKind is a tuple of Group Kind and Version.
+type ParamKind struct {
+	// APIVersion is the API group version the resources belong to.
+	// In format of "group/version".
+	// Required.
+	APIVersion string
+
+	// Kind is the API kind the resources belong to.
+	// Required.
+	Kind string
+}
+
+// Validation specifies the CEL expression which is used to apply the validation.
+type Validation struct {
+	// Expression represents the expression which will be evaluated by CEL.
+	// ref: https://github.com/google/cel-spec
+	// CEL expressions have access to the contents of the Admission request/response, organized into CEL variables as well as some other useful variables:
+	//
+	//'object' - The object from the incoming request. The value is null for DELETE requests.
+	//'oldObject' - The existing object. The value is null for CREATE requests.
+	//'request' - Attributes of the admission request([ref](/pkg/apis/admission/types.go#AdmissionRequest)).
+	//'params' - Parameter resource referred to by the policy binding being evaluated. Only populated if the policy has a ParamSource.
```

Comment:
ParamKind

Reply:
Same as another. Will resolve this one and discuss on another comment: https://github.com/kubernetes/kubernetes/pull/112883#discussion_r1003749276